README


This archive contains the distribution Redis,
version 1.975:

  Perl binding for Redis database

This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)




Changes
Revision history for Redis

1.975     2014-08-03 20:50:25 Europe/Amsterdam

   * Clarification about reconnect and read_timeout (#89)
   * Test::CPAN::Meta (et al.) are required even if unused (#92)

1.974     2014-05-16 21:42:48 Europe/Amsterdam

   * released as stable version, no change

1.973_04  2014-05-12 22:53:06 Europe/Amsterdam

   * release again, last one was screwed up.
   * fix #85 (PR #86) reconnect during transaction

1.973_03  2014-05-12 22:49:07 Europe/Amsterdam

   * fix #85 (PR #86) reconnect during transaction

1.973_02  2014-04-30 12:04:29 Europe/Amsterdam

   * merge PR #84 optimize try read sock

1.973_01  2014-04-26 18:00:31 Europe/Amsterdam

   * use new network code from Ivan Kruglov
   * fix sentinel tests
   * fix #81: doc for 'every' option

1.972     2014-02-18 00:54:01 Europe/Amsterdam
   * Sentinel features (connections, timeouts, etc) support
   * various bugfixes and testfixes
   * fix network code for BSDs
   * no_auto_connect_on_new

1.971     2014-02-01 09:55:11 Europe/Paris

   * skip some tests that fail on some platforms for now

1.970     2014-01-30 15:07:42 Europe/Amsterdam

   * fix tests breaking in some case

1.969     2014-01-30 13:19:28 Europe/Amsterdam

    * Clarification for (p)unsubscribe commands.
    * use Test::TCP for testing

1.968     2014-01-30 12:19:11 Europe/Amsterdam

    * Add a no_auto_connect_on_new parameter to new() to allow users
      to call $x = Redis->new and then $x->connect, instead of Redis
      auto-connecting. Useful for tuning the cnx_timeout parameter.

1.967     2013-12-28 22:58:55 Europe/Paris
    * use new IO::Socket::Timeout with different API

1.966     2013-12-17 13:58:33 Europe/Amsterdam
    * fix tests for Redis 2.8

1.965     2013-11-29 09:28:36 Europe/Amsterdam

    * fix #60: TEST_REQUIRES needs newer MakeMaker
    * fix #34: perl-redis doesn't receive subcribed message if server restart 
    * fix #38: select new database doesn't survive after reconnect
    * minor documentation fixes

1.964     2013-11-14 15:45:08 Europe/Amsterdam
    * minor fix to dist.ini to fix github url

1.963     2013-11-13 22:44:29 Europe/Paris

    * Add documentation for the slowlog command PR #44 (rgs)
    * doc fix, duplicate spop PR #48 (nightlord)
    * Redis client is now fork safe PR #51 (songmu)
    * rewrite dist.ini

1.962     2013-10-16T14:58:30Z
    * merge #54 (fix for broken test)
    * Parameter name to new() can be a CodeRef, dynamic connection names
    * admin: added co-maintainer Damien Krotkine <dams@zarb.org>

1.961     2013-01-23T15:09:47Z
    * Add wait_one_response() for better pipeline management

1.960     2013-01-23T11:28:40Z
    * Make the new() name parameter safe to use with all redis-
      server versions

1.959     2013-01-22T14:46:42Z
    * __try_read_sock: test error conditions more thoroughly (issue #31)
    * Improve Test::SpawnRedisServer: multi-server support
    * tidyall the code and docs

1.958     2013-01-15T16:54:40Z
    * Support for name constructor parameter, set connection name
    * Add documentation for CLIENT * commands
    * Improve reliability of Test::SpawnRedisServer

1.957     2013-01-15T13:18:07Z
    * Fix minimum Perl version

1.956     2013-01-15T10:35:10Z
    * Add on_connect callback to Redis constructor (issue 28)
    * Make sure quit() doesn't die when the socket is already dead (issue 30);
    * Switch to Digest::SHA, one less non-core dep RT#81841
    * Try and make Travis-CI more useful, ask for a redis-server
    * Update SUPPORT section, moving bugtracker to GitHub issues

1.955     2012-10-10T11:43:44Z
    * Skip 30-scripts.t if the redis-server found lacks script support

1.954     2012-10-10T11:16:22Z
    * Support for multi-word commands such as "SCRIPT LOAD".
    * Try another fix for Windows non-blocking reads

1.953     2012-09-05T00:49:11Z
    * Tweak travis.ci setup

1.952     2012-09-04T11:22:18Z
    * Added automatic authentication after connection establishment
    * Support Redis 2.6: Aaron Crane <arc@cpan.org>
    * Attempt to fix non-blocking read on Windows
    * Enable travis.ci support on the repository

1.951     2012-03-13T10:17:09Z
    * Remove Tie::StdHash from our dependencies list, its part of core
      and it creates problems with some CPAN clients

1.950     2012-03-12T13:54:10Z
    * DEPRECATED: the encoding attribute to new() - we will not support
      automatic decoding after 2.000, so please test your code with
      encoding => undef now.
    * Add pipeline support: Aaron Crane arc@github++!
    * Cache AUTOLOAD calls: improves performance a bit, fixes #2
    * Fix: apply reconnect logic to KEYS and INFO
    * Fix: forbid PING and SHUTDOWN in SUBSCRIBE mode
    * Updated docs covering pipelining and multi/exec
    * Updated docs to point users to Github for code and issues

1.926 Wed Jan 11 15:48:11 UTC 2012
    * Fix auto-reconnect when the server is dead or a
      client timeout.
      See https://github.com/melo/perl-redis/issues/1#issuecomment-3444989

1.925 Tue Jan 10 16:02:04 UTC 2012
    * Implemented auto-reconnect
    * Add support for UNIX domain sockets
    * Make REDIS_SERVER work with both TCP and UNIX domain sockets
    * Make the test suite workaround a missing redis-server binary
    * Assorted small bug fixes
    * Improve documentation

1.904 Sat Mar  5 23:10:48 UTC 2011
    * Fix bug with NIL multi-bulk replies (Case 42) RT#64040

1.903 Tue Feb 22 13:04:24 UTC 2011
    * remove the Guard dependency

1.902 Sat Feb  5 12:38:57 UTC 2011
    * fix: ping() no longer dies (RT #62489)
    * fix: shutdown() no longer dies

1.901 Sat Feb  5 11:15:04 UTC 2011
    * Released 1.900_01 as latest version

1.900_01  Sun Jan 30 06:03:14 UTC 2011
    * admin: change of maintainer to Pedro Melo <melo@simplicidade.org>
    * feature: full support for Redis 2.x multi-bulk protocol
    * feature: support for Redis PUBLISH/SUBSCRIBE commands

t/tlib/Test/SpawnRedisServer.pm
#
# This file is part of Redis
#
# This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.
#
# This is free software, licensed under:
#
#   The Artistic License 2.0 (GPL Compatible)
#
package    # Hide from PAUSE
  Test::SpawnRedisServer;

use strict;
use warnings;
use Redis;
use File::Temp;
use IPC::Cmd qw(can_run);
use POSIX ":sys_wait_h";
use base qw( Exporter );

use Net::EmptyPort qw(empty_port);

our @EXPORT    = qw( redis );
our @EXPORT_OK = qw( redis reap );

sub redis {
  my %params = (
    timeout => 120,
    @_,
  );

  my ($fh, $fn) = File::Temp::tempfile();

  my $port = empty_port();

  my $local_port = $port;
  $params{port}
    and $local_port = $params{port};

  my $addr = "127.0.0.1:$local_port";

  unlink("redis-server-$addr.log");
  unlink('dump.rdb');

  $fh->print("
    timeout $params{timeout}
    appendonly no
    daemonize no
    port $local_port
    bind 127.0.0.1
    loglevel debug
    logfile redis-server-$addr.log
  ");
  $fh->flush;

  Test::More::diag("Spawn Redis at $addr, cfg $fn") if $ENV{REDIS_DEBUG};

  my $redis_server_path = $ENV{REDIS_SERVER_PATH} || 'redis-server';
  if (!can_run($redis_server_path)) {
    Test::More::plan skip_all => "Could not find binary redis-server";
    return;
  }

  my ($ver, $c);
  eval { ($ver, $c) = spawn_server($redis_server_path, $fn, $addr) };
  if (my $e = $@) {
    reap();
    Test::More::plan skip_all => "Could not start redis-server: $@";
    return;
  }

  if (my $rvs = $params{requires_version}) {
    if (!defined $ver) {
      $c->();
      Test::More::plan skip_all => "This tests require at least redis-server $rvs, could not determine server version";
      return;
    }

    my ($v1, $v2, $v3) = split(/[.]/, $ver);
    my ($r1, $r2, $r3) = split(/[.]/, $rvs);
    if ($v1 < $r1 or $v1 == $r1 and $v2 < $r2 or $v1 == $r1 and $v2 == $r2 and $v3 < $r3) {
      $c->();
      Test::More::plan skip_all => "This tests require at least redis-server $rvs, server found is $ver";
      return;
    }
  }

  return ($c, $addr, $ver, split(/[.]/, $ver), $local_port);
}

sub spawn_server {
  my $addr = pop;
  my $pid  = fork();
  if ($pid) {    ## Parent
    require Test::More;
    Test::More::diag("Starting server with pid $pid") if $ENV{REDIS_DEBUG};

    my $redis   = Redis->new(server => $addr, reconnect => 5, every => 200);
    my $version = $redis->info->{redis_version};
    my $alive   = $$;

    my $c = sub {
      return unless $alive;
      return unless $$ == $alive;    ## only our creator can kill us

      Test::More::diag("Killing server at $pid") if $ENV{REDIS_DEBUG};
      kill(15, $pid);

      my $failed = reap($pid);
      Test::More::diag("Failed to kill server at $pid")
        if $ENV{REDIS_DEBUG} and $failed;
      unlink("redis-server-$addr.log");
      unlink('dump.rdb');
      $alive = 0;

      return !$failed;
    };

    return $version => $c;
  }
  elsif (defined $pid) {    ## Child
    exec(@_);
    warn "## In child Failed exec of '@_': $!, ";
    exit(1);
  }

  die "Could not fork(): $!";
}

sub reap {
  my ($pid, $limit) = @_;
  $pid   = -1 unless $pid;
  $limit = 3  unless $limit;

  my $try = 0;
  local $?;
  while ($try++ < $limit) {
    my $ok = waitpid($pid, WNOHANG);
    $try = 0, last if $ok > 0;
    sleep(1);
  }

  return $try;
}

1;


t/tlib/Test/SpawnRedisTimeoutServer.pm
#
# This file is part of Redis
#
# This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.
#
# This is free software, licensed under:
#
#   The Artistic License 2.0 (GPL Compatible)
#
package    # Hide from PAUSE
  Test::SpawnRedisTimeoutServer;

use strict;
use warnings;
use Test::TCP;

sub create_server_with_timeout {
    my $timeout = shift;

    Test::TCP->new(
        code => sub {
            my $port   = shift;
            my $socket = IO::Socket::INET->new(
                Listen    => 5,
                Timeout   => 1,
                Reuse     => 1,
                Blocking  => 1,
                LocalPort => $port
            ) or die "ops $!";

            my $buffer;
            while (1) {
                my $client = $socket->accept();
                if (defined (my $got = <$client>)) {
                    sleep $timeout;
                    $client->print("+42\r\n");
                }

                # sleep($in_timeout) if $in_timeout;
                # if ( $client->sysread( $buffer, 5 ) ) {
                #     my $response =
                #       pack( 'N a*', bytes::length($message), $message );

                #     $client->syswrite($response);
                #     sleep(1);
                # }

            }
        },
    );
}
1;


lib/Redis/Hash.pm
#
# This file is part of Redis
#
# This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.
#
# This is free software, licensed under:
#
#   The Artistic License 2.0 (GPL Compatible)
#
package Redis::Hash;
{
  $Redis::Hash::VERSION = '1.974';
}

# ABSTRACT: tie Perl hashes to Redis hashes
# VERSION
# AUTHORITY

use strict;
use warnings;
use Tie::Hash;
use base qw/Redis Tie::StdHash/;


sub TIEHASH {
  my ($class, $prefix, @rest) = @_;
  my $self = $class->new(@rest);

  $self->{prefix} = $prefix ? "$prefix:" : '';

  return $self;
}

sub STORE {
  my ($self, $key, $value) = @_;
  $self->set($self->{prefix} . $key, $value);
}

sub FETCH {
  my ($self, $key) = @_;
  $self->get($self->{prefix} . $key);
}

sub FIRSTKEY {
  my $self = shift;
  $self->{prefix_keys} = [$self->keys($self->{prefix} . '*')];
  $self->NEXTKEY;
}

sub NEXTKEY {
  my $self = shift;

  my $key = shift @{ $self->{prefix_keys} };
  return unless defined $key;

  my $p = $self->{prefix};
  $key =~ s/^$p// if $p;
  return $key;
}

sub EXISTS {
  my ($self, $key) = @_;
  $self->exists($self->{prefix} . $key);
}

sub DELETE {
  my ($self, $key) = @_;
  $self->del($self->{prefix} . $key);
}

sub CLEAR {
  my ($self) = @_;
  $self->del($_) for $self->keys($self->{prefix} . '*');
  $self->{prefix_keys} = [];
}


1;    ## End of Redis::Hash

__END__

=pod

=encoding UTF-8

=head1 NAME

Redis::Hash - tie Perl hashes to Redis hashes

=head1 VERSION

version 1.974

=head1 DESCRIPTION

Ties a Perl hash to Redis. Note that it doesn't use Redis Hashes, but
implements a fake hash using regular keys like "prefix:KEY".

If no C<prefix> is given, it will tie the entire Redis database as a hash.

Future versions will also allow you to use real Redis hash structures.

=head1 SYNOPSYS

    ## Create fake hash using keys like 'hash_prefix:KEY'
    tie %my_hash, 'Redis::Hash', 'hash_prefix', @Redis_new_parameters;

    ## Treat the entire Redis database as a hash
    tie %my_hash, 'Redis::Hash', undef, @Redis_new_parameters;

    $value = $my_hash{$key};
    $my_hash{$key} = $value;

    @keys   = keys %my_hash;
    @values = values %my_hash;

    %my_hash = reverse %my_hash;

    %my_hash = ();

=head1 AUTHORS

=over 4

=item *

Pedro Melo <melo@cpan.org>

=item *

Damien Krotkine <dams@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


lib/Redis/List.pm
#
# This file is part of Redis
#
# This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.
#
# This is free software, licensed under:
#
#   The Artistic License 2.0 (GPL Compatible)
#
package Redis::List;
{
  $Redis::List::VERSION = '1.974';
}

# ABSTRACT: tie Perl arrays to Redis lists
# VERSION
# AUTHORITY

use strict;
use warnings;
use base qw/Redis Tie::Array/;


sub TIEARRAY {
  my ($class, $list, @rest) = @_;
  my $self = $class->new(@rest);

  $self->{list} = $list;

  return $self;
}

sub FETCH {
  my ($self, $index) = @_;
  $self->lindex($self->{list}, $index);
}

sub FETCHSIZE {
  my ($self) = @_;
  $self->llen($self->{list});
}

sub STORE {
  my ($self, $index, $value) = @_;
  $self->lset($self->{list}, $index, $value);
}

sub STORESIZE {
  my ($self, $count) = @_;
  $self->ltrim($self->{list}, 0, $count);

#		if $count > $self->FETCHSIZE;
}

sub CLEAR {
  my ($self) = @_;
  $self->del($self->{list});
}

sub PUSH {
  my $self = shift;
  my $list = $self->{list};

  $self->rpush($list, $_) for @_;
}

sub POP {
  my $self = shift;
  $self->rpop($self->{list});
}

sub SHIFT {
  my ($self) = @_;
  $self->lpop($self->{list});
}

sub UNSHIFT {
  my $self = shift;
  my $list = $self->{list};

  $self->lpush($list, $_) for @_;
}

sub SPLICE {
  my ($self, $offset, $length) = @_;
  $self->lrange($self->{list}, $offset, $length);

  # FIXME rest of @_ ?
}

sub EXTEND {
  my ($self, $count) = @_;
  $self->rpush($self->{list}, '') for ($self->FETCHSIZE .. ($count - 1));
}

sub DESTROY { $_[0]->quit }

1;    ## End of Redis::List

__END__

=pod

=encoding UTF-8

=head1 NAME

Redis::List - tie Perl arrays to Redis lists

=head1 VERSION

version 1.974

=head1 SYNOPSYS

    tie @my_list, 'Redis::List', 'list_name', @Redis_new_parameters;

    $value = $my_list[$index];
    $my_list[$index] = $value;

    $count = @my_list;

    push @my_list, 'values';
    $value = pop @my_list;
    unshift @my_list, 'values';
    $value = shift @my_list;

    ## NOTE: fourth parameter of splice is *NOT* supported for now
    @other_list = splice(@my_list, 2, 3);

    @my_list = ();

=head1 AUTHORS

=over 4

=item *

Pedro Melo <melo@cpan.org>

=item *

Damien Krotkine <dams@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


lib/Redis/Sentinel.pm
#
# This file is part of Redis
#
# This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.
#
# This is free software, licensed under:
#
#   The Artistic License 2.0 (GPL Compatible)
#
package Redis::Sentinel;
{
  $Redis::Sentinel::VERSION = '1.974';
}

# ABSTRACT: Redis Sentinel interface

use warnings;
use strict;

use Carp;

use base qw(Redis);

sub new {
    my ($class, %args) = @_;
    # these args are not allowed when contacting a sentinel
    delete @args{qw(sentinels service)};

    $class->SUPER::new(%args);
}

sub get_service_address {
    my ($self, $service) = @_;
    my ($ip, $port) = $self->sentinel('get-master-addr-by-name', $service);
    defined $ip
      or return;
    $ip eq 'IDONTKNOW'
      and return $ip;
    return "$ip:$port";
}

sub get_masters {
    map { +{ @$_ }; } @{ shift->sentinel('masters') || [] };
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Redis::Sentinel - Redis Sentinel interface

=head1 VERSION

version 1.974

=head1 SYNOPSIS

    my $sentinel = Redis::Sentinel->new( ... );
    my $service_address = $sentinel->get_service_address('mymaster');
    my @masters = $sentinel->get_masters;

=head1 DESCRIPTION

This is a subclass of the Redis module, specialized into connecting to a
Sentinel instance. Inherits from the C<Redis> package;

=head1 CONSTRUCTOR

=head2 new

See C<new> in L<Redis.pm>. All parameters are supported, except C<sentinels>
and C<service>, which are silently ignored.

=head1 METHODS

All the methods of the C<Redis> package are supported, plus the aditional following methods:

=head2 get_service_address

Takes the name of a service as parameter, and returns either void (emptly list)
if the master couldn't be found, the string 'IDONTKNOW' if the service is in
the sentinel config but cannot be reached, or the string C<"$ip:$port"> if the
service were found.

=head2 get_masters

Returns a list of HashRefs representing all the master redis instances that
this sentinel monitors.

=head1 AUTHORS

=over 4

=item *

Pedro Melo <melo@cpan.org>

=item *

Damien Krotkine <dams@cpan.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Pedro Melo, Damien Krotkine.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Makefile.PL
# This Makefile.PL for Redis was generated by
# Dist::Zilla::Plugin::MakeMaker::Fallback 0.005
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.19.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

BEGIN {
my %configure_requires = (
    'Module::Build::Tiny' => '0.036',
    'ExtUtils::MakeMaker' => '6.63_03',
);

my @missing = grep {
    ! eval "require $_; $_->VERSION($configure_requires{$_}); 1"
} keys %configure_requires;

if (not @missing)
{
    print "Congratulations, your toolchain understands 'configure_requires'!\n\n";
}
else
{
    $ENV{PERL_MM_FALLBACK_SILENCE_WARNING} or warn <<'EOW';
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

If you're installing manually, please retrain your fingers to run Build.PL
when present instead.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are doing something overly clever, in which case you
should consider setting the 'prefer_installer' config option in CPAN.pm, or
'prefer_makefile' in CPANPLUS, to 'mb" and '0' respectively.

You can also silence this warning for future installations by setting the
PERL_MM_FALLBACK_SILENCE_WARNING environment variable.

EOW

    sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
}
}

use 5.008;

use ExtUtils::MakeMaker 6.63_03;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl binding for Redis database",
  "AUTHOR" => "Pedro Melo <melo\@cpan.org>, Damien Krotkine <dams\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.63_03",
    "Module::Build::Tiny" => "0.036"
  },
  "DISTNAME" => "Redis",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Redis",
  "PREREQ_PM" => {
    "IO::Socket::Timeout" => "0.22",
    "Try::Tiny" => 0
  },
  "TEST_REQUIRES" => {
    "Digest::SHA" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IO::String" => 0,
    "IPC::Cmd" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.98",
    "Test::SharedFork" => 0,
    "Test::TCP" => "1.19"
  },
  "VERSION" => "1.975",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Digest::SHA" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IO::Socket::Timeout" => "0.22",
  "IO::String" => 0,
  "IPC::Cmd" => 0,
  "IPC::Open3" => 0,
  "Test::Deep" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.98",
  "Test::SharedFork" => 0,
  "Test::TCP" => "1.19",
  "Try::Tiny" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



